File: algorithms/2_medium/M/02874_maximum_value_of_an_ordered_triplet_ii/solution.cpp
class Solution {
public:
    static long long maximumTripletValue(const vector<int>& nums) {
        const size_t len = nums.size();
        if (len < 3) {
            cerr << "invalid input";
            exit(1);
        }

        int diff = nums[0] - nums[1];
        int max = std::max(nums[0], nums[1]);
        int64_t res = 0;

        for (int i = 2; i < len; i += 1) {
            res = std::max(res, static_cast<int64_t>(diff) * nums[i]);
            diff = std::max(diff, max - nums[i]);
            max = std::max(max, nums[i]);
        }

        return res;
    }
};

File: algorithms/2_medium/M/02874_maximum_value_of_an_ordered_triplet_ii/solution.rs
impl Solution {
    pub fn maximum_triplet_value(nums: Vec<i32>) -> i64 {
        let len = nums.len();
        assert!(len >= 3);
        let mut diff = nums[0] - nums[1];
        let mut max = nums[0].max(nums[1]);
        let mut res = i64::MIN;

        for i in 2..len {
            res = res.max(diff as i64 * nums[i] as i64);
            diff = diff.max(max - nums[i]);
            max = max.max(nums[i]);
        }

        res.max(0)
    }
}
